[AMDGPU] Fix undefined scc register in successor block of SI_KILL terminators

mssefat opened the pull request:
Fix issue 131298 where an undefined $scc register causes verifier errors
when using SI_KILL_F32_COND_IMM_TERMINATOR instructions. The problem occurs
because the $scc register defined in a comparison before the kill terminator
is used in successor blocks, but was not properly marked as live-in.

This patch:
- Adds code to check if SCC is used in the successor block
- Adds SCC as a live-in to successor blocks
- Handles both explicit and implicit uses of SCC

With this patch the machine verifier no longer reports undefined $scc errors in
following kill terminator instruction.

Fixes #131298

mssefat committed:
[AMDGPU] Fix undefined $scc in successor blocks of SI_KILL terminators

Fixes #131298

Fix issue 131298 where an undefined $scc register causes verifier errors
when using SI_KILL_F32_COND_IMM_TERMINATOR instructions. The problem occurs
because the $scc register defined in a comparison before the kill terminator
is used in successor blocks, but was not properly marked as live-in.

This patch:
- Adds code to check if SCC is used in the successor block
- Adds SCC as a live-in to successor blocks
- Handles both explicit and implicit uses of SCC

With this patch the machine verifier no longer reports undefined $scc errors in
following kill terminator instruction.

mssefat committed:
[AMDGPU] Fix undefined $scc in successor blocks of SI_KILL terminators

Fixes #131298

Fix issue 131298 where an undefined $scc register causes verifier errors when using SI_KILL_F32_COND_IMM_TERMINATOR instructions. The problem occurs because the $scc register defined in a comparison before the kill terminator is used in successor blocks, but was not properly marked as live-in.

This patch:
- Adds code to check if SCC is used in the successor block
- Adds SCC as a live-in to successor blocks
- Handles both explicit and implicit uses of SCC

With this patch the machine verifier no longer reports undefined $scc errors in following kill terminator instruction.

jmmartinez commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -4516,6 +4516,38 @@ SITargetLowering::splitKillBlock(MachineInstr &MI,
   MachineBasicBlock *SplitBB = BB->splitAt(MI, false /*UpdateLiveIns*/);
```
What if we set `UpdateLiveIns` to true ? Wouldn't that cover the case we're trying to cover here ?

mssefat replied:
I've pushed the updates, where UpdateLiveIns is now set to true. To support this change, I moved the call to finalizeLowering to the beginning of FinalizeISel::runImpl. This ensures that reserved registers are frozen before live-ins are added. Without this change, llvm::addLiveIns() triggers an assertion failure due to the check for reservedRegsFrozen() when reserved registers are not yet finalized. Please review.

jmmartinez commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -4516,6 +4516,38 @@ SITargetLowering::splitKillBlock(MachineInstr &MI,
   MI.setDesc(TII->getKillTerminatorFromPseudo(MI.getOpcode()));
+
+  // Check if SCC register is used in the successor block
+  bool IsSCCUsedInSuccessor = false;
+  for (const MachineInstr &SuccMI : *SplitBB) {
+    // Check for uses of SCC in the instruction's operands
+    for (const MachineOperand &MO : SuccMI.operands()) {
+      if (MO.isReg() && MO.getReg() == AMDGPU::SCC && !MO.isDef()) {
+        IsSCCUsedInSuccessor = true;
+        break;
+      }
+    }
+
+    // Also check for implicit uses of SCC
+    if(!IsSCCUsedInSuccessor){
+      const MCInstrDesc &Desc = SuccMI.getDesc();
+      if (Desc.hasImplicitUseOfPhysReg(AMDGPU::SCC)) {
+        IsSCCUsedInSuccessor = true;
+        break;
+      }
+    }
+    
+    if (IsSCCUsedInSuccessor)
+      break;
+  }
```
If you move this code to a helper function we would end up with less awkward branching:

```cpp
for (const MachineInstr &SuccMI : *SplitBB) {
    // Check for uses of SCC in the instruction's operands
    for (const MachineOperand &MO : SuccMI.operands())
      if (MO.isReg() && MO.getReg() == AMDGPU::SCC && !MO.isDef())
        return true;

    // Also check for implicit uses of SCC
    const MCInstrDesc &Desc = SuccMI.getDesc();
    if (Desc.hasImplicitUseOfPhysReg(AMDGPU::SCC))
      return true;
}
```

arsenm reviewed (commented):
Should also include MIR test. We shouldn't really need to scan the entire successor for this

mssefat committed:
[AMDGPU] Fix undefined scc register in successor block of SI_KILL terminators

Fixes #131298

This patch addresses issue #131298, where the use of SI_KILL_F32_COND_IMM_TERMINATOR instructions leads to verifier errors due to an undefined $scc register. The error occurs when $scc, defined by a comparison instruction prior to the kill terminator, is used in successor blocks without being correctly marked as a live-in.

To fix this:

The call to finalizeLowering is moved to the beginning of FinalizeISel::runImpl to ensure reserved registers are frozen before live-ins are added.

In SITargetLowering::splitKillBlock, the splitAt function is now called with UpdateLiveIns = true to properly update live-in registers.

With these changes, the machine verifier no longer reports undefined $scc errors for kill terminator instructions.

mssefat commented:
> Should also include MIR test. We shouldn't really need to scan the entire successor for this

I have pushed the updates. Please review.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -4513,7 +4513,7 @@ Register SITargetLowering::getRegisterByName(const char *RegName, LLT VT,
 SITargetLowering::splitKillBlock(MachineInstr &MI,
                                  MachineBasicBlock *BB) const {
-  MachineBasicBlock *SplitBB = BB->splitAt(MI, false /*UpdateLiveIns*/);
+  MachineBasicBlock *SplitBB = BB->splitAt(MI, true /*UpdateLiveIns*/);
```
```suggestion
  MachineBasicBlock *SplitBB = BB->splitAt(MI, /*UpdateLiveIns=*/true);
```

arsenm commented on llvm/lib/CodeGen/FinalizeISel.cpp:
```diff
@@ -47,6 +47,8 @@ static std::pair<bool, bool> runImpl(MachineFunction &MF) {
   const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
   const TargetLowering *TLI = MF.getSubtarget().getTargetLowering();
 
+  TLI->finalizeLowering(MF);
```
Do you really need this reordering? I'd expect this to be called after the custom insertions

mssefat replied:
I think we need this reordering. Without this reordering, llvm::addLiveIns() invoked from MachineBasicBlock::splitAt() triggers an assertion failure due to the check for reservedRegsFrozen() when reserved registers are not yet finalized.

arsenm replied:
This API is bad. We probably should split these into separate steps, but given no tests apparently rely on this order I guess this is fine for now

mssefat committed:
Update llvm/lib/Target/AMDGPU/SIISelLowering.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>
